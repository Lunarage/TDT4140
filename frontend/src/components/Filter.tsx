import React, { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import styled from 'styled-components';
import { redHexColorHover } from '../consts';
import { getEvents, getOrgs } from '../store/actionCreators';
import { State } from '../store/types';
import { CustomButton, TextWrapper } from './Button';
import Button from './Button';
import './DropDown.css';

const FilterWrapper = styled.div`
  background-color: #c91801;
  height: 962px;
  width: inherit;
  border-style: none;
  display: flex;
  flex-direction: column;
  align-items: left;
  justify-content: flex-start;
  align-content: space-between;
  `;

const FilterHeader = styled.h2`
  font-weight: bold;
  font-size: 21px;
  color: white;
  text-shadow: 4px 4px 4px darkred;
  padding: 5% 2% 0% 4%;
  `;

const FilterSearchInput = styled.div`
  width: 100% inherit;
  height: 30px inherit;
  background-color: #ffc6c6;
  margin: 0 4% 4% 4%;
  padding: 2% 2% 2% 3%;
  vertical-align: middle;
  color: #313131;
`;

const CheckBoxWrap = styled.div`
  display: flex;
  flex-direction: row;
  align-items: center;
  padding-left: 4%;
  margin: 4% 4% 4% 0;
`;

const CheckBox = styled.div`
  width: 20px;
  height: 20px;
  border-style: solid;
  border-color: white;
  opacity: 1;
  box-shadow: 4px 4px 4px #8b0000;

  &:hover {
    background-color: white;
  }
`;

const CheckBoxTittel = styled.div`
  font-weight: 600;
  font-size: 17px;
  color: white;
  padding-left: 4%;
  text-shadow: 4px 4px 4px #8b0000;
`;

const DropDownWrapAll = styled.div`
  height: fit-content;
  margin-top: 10%;
`;

const SelectedFiltersWrap = styled.div`
  background-color: rgba(255,78,55,0.31);
  width = 100% inherit;
  height: inherit;
  margin: 40px 4% 4% 4%;
  padding = 2%;
  flex-wrap: wrap;
  justify-content: flex-start;
  color: white;
`;

const FilterButton = styled(CustomButton) `
  background-color: white;
  width: fit-content;
  min-width: 5em;
`

const ButtonTextWrapper = styled(TextWrapper) `
  color: darkred;
`


const SearchField = () => {
  return (
    <FilterSearchInput>
      Skriv inn nøkkelord ...
    </FilterSearchInput>
  )
}

interface CheckBoxProps {
  tittel: string;
  onClick: () => void;
}

const CheckBoxes = (props: CheckBoxProps) => {
  return (
    <CheckBoxWrap>
      <CheckBox onClick= {props.onClick}/>
      <CheckBoxTittel>{props.tittel}</CheckBoxTittel>
    </CheckBoxWrap>
  );
}


interface DropDownProps {
  tittel: string;
}

const DropDown = (props: DropDownProps) => {
  
  return (
    <form>
      <label className="dropDownLabel">Kategorier</label>
      <select className="dropDownWrapper">
          <option className="dropDownText">Alt test 1</option>
          <option className="dropDownText">Alt test 2</option>
        </select>
    </form>
    
  )
}


interface SelectedFiltersProps {
  tittel: string;
  filters: Array<string>;
}

const SelectedFilters = ( props: SelectedFiltersProps) => {
  //const test = ["Innendørs", "Sykkel", "Skistøvler", "Helikopter"];
  const renderButtons =  (filter: string) => {
    return (
      <FilterButton>
        <ButtonTextWrapper>{filter}</ButtonTextWrapper>
      </FilterButton>
    );
  }

  return (
    <SelectedFiltersWrap>
      <p>{props.tittel}</p>
      {props.filters.map(renderButtons)}
    </SelectedFiltersWrap>
  );
}


const Filter = () => {
  const dispatch = useDispatch();
  const [selectedFilters, setSelectedFilters] = useState<string[]>(["Det funker, yay!"]);
  const [urlFilters, setUrlFilters] = useState<string>("");  

  const {
    categories: categoriesData,
    isLoading: categoriesLoading,
  } = useSelector((state: State) => state.categoriesReducer);
  console.log(categoriesData);

  const {
    equipment: equipmentData,
    isLoading: equipmentLoading,
  } = useSelector((state: State) => state.equipmentReducer);

  const {
    organizations,
    isLoading: orgsLoading,
  } = useSelector((state: State) => state.orgsReducer);
  console.log(organizations);

  useEffect(() => {
    if (!organizations) {
      dispatch(getOrgs());
    }
  }, [dispatch, organizations]);

  useEffect(() => {
    dispatch(getEvents(urlFilters));
  }, [dispatch, urlFilters]);


  const handleOnClickSelect = (filter: string) => {
    setSelectedFilters(selectedFilters.concat(filter));
  }

  const handleClear = () => {
    setSelectedFilters([]);
  }


  return (
    <FilterWrapper>
      <FilterHeader> Søk etter aktiviteter: </FilterHeader>
      <SearchField />
      <CheckBoxes tittel="Innendørs" onClick={() => handleOnClickSelect("Innendørs")}/>
      <CheckBoxes tittel="Utendørs" onClick={() => handleOnClickSelect("Utendørs")}/>
      <CheckBoxes tittel="Gratis (Pris: 0kr)" onClick={() => handleOnClickSelect("Gratis (Pris: 0kr)")}/>
      <DropDownWrapAll>
        <DropDown tittel="Kategori" />
        <DropDown tittel="Utstyr" />
        <DropDown tittel="Organisasjon" />
        <DropDown tittel="Intensistet" />
      </DropDownWrapAll>
      <SelectedFilters tittel="Utvalgte filtre: " filters={selectedFilters} />
      <Button text={"Klikk for å nulstille filtre"} onClickFunc={handleClear}/>
    </FilterWrapper>
  );
  }


export default Filter;
