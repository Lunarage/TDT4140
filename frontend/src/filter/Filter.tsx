import { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import styled from 'styled-components';
import { getCategories, getEquipment, getEvents, getOrgs } from '../store/actionCreators';
import { State } from '../store/types';
import { CustomButton, TextWrapper } from '../components/Button';
import Button from '../components/Button';
import DropDown from './DropDown';
import SearchField from './SearchField';

// styling for tha main filter component
const FilterWrapper = styled.div`
  background-color: #c91801;
  height: 962px;
  width: inherit;
  border-style: none;
  display: flex;
  flex-direction: column;
  align-items: left;
  justify-content: flex-start;
  align-content: space-between;
  `;

// styling for the header of 
const FilterHeader = styled.h2`
  font-weight: bold;
  font-size: 21px;
  color: white;
  text-shadow: 4px 4px 4px darkred;
  padding: 5% 2% 0% 4%;
`;

// size wrapping
const DropDownWrapAll = styled.div`
  height: fit-content;
  margin-top: 10%;
`;

// styling for container that displays buttons with information about selected filters
const SelectedFiltersWrap = styled.div`
  background-color: rgba(255,78,55,0.31);
  width = 100% inherit;
  height: inherit;
  margin: 40px 4% 4% 4%;
  padding = 2%;
  flex-wrap: wrap;
  justify-content: flex-start;
  color: white;
`;

// styling for button displaying a selected filter
const FilterButton = styled(CustomButton)`
  background-color: white;
  width: fit-content;
  min-width: 5em;
`

// styling of text of FilterButton
const ButtonTextWrapper = styled(TextWrapper)`
  color: darkred;
`

interface SelectedFiltersProps {
  tittel: string;
  filters: string[];
}

// component that shows/displays filters that have been selected and that is currently being searched on.
const SelectedFilters = (props: SelectedFiltersProps) => {
  const renderButtons = (filter: string) => {
    return (
      <FilterButton key={filter}>
        <ButtonTextWrapper>{filter}</ButtonTextWrapper>
      </FilterButton>
    );
  }
  return (
    <SelectedFiltersWrap>
      <p>{props.tittel}</p>
      {props.filters.map(renderButtons)}
    </SelectedFiltersWrap>
  );
}

const availableIntensity = ["Intensitet = 1", "Intensitet = 2", "Intensitet = 3", "Intensitet = 4", "Intensitet = 5"];

// Main function for the filter component
const Filter = () => {
  const dispatch = useDispatch();

  // States of data from database
  const [availableOrganizations, setAvailableOrganizations] = useState<string[]>([]);
  const [availableCategories, setAvailableCategories] = useState<string[]>([]);
  const [availableEquipment, setAvailableEquipment] = useState<string[]>([]);

  //States of the filter component
  const [selectedFilters, setSelectedFilters] = useState<string[]>([]);  // Collects all selected filters// Collects filters chosen by checkboxes
  const [selectedEquipment, setSelectedEquipment] = useState<string>("");
  const [selectedCategory, setSelectedCategory] = useState<string>("");
  const [selectedOrganization, setSelectedOrganization] = useState<string>("");
  const [selectedIntensity, setSelectedIntensity] = useState<string>("");
  const [selectedKeyWord, setSelectedKeyWord] = useState<string>("");
  const [selectedLocation, setSelectedLocation] = useState<string>("");


  const {
    categories: categoriesData
  } = useSelector((state: State) => state.categoriesReducer);


  const {
    equipment: equipmentData
  } = useSelector((state: State) => state.equipmentReducer);

  const {
    organizations
  } = useSelector((state: State) => state.orgsReducer);


  useEffect(() => {
    if (!organizations) {
      dispatch(getOrgs());
    } else {
      let orgs: string[] = [];
      organizations.forEach((org) => { orgs.push(org.name) })
      setAvailableOrganizations(orgs);
    }
  }, [dispatch, organizations]);


  useEffect(() => {
    if (!categoriesData) {
      dispatch(getCategories());
    } else {
      let cats: string[] = [];
      categoriesData.forEach((cat) => { cats.push(cat.name) })
      setAvailableCategories(cats);
    }
  }, [dispatch, categoriesData]);


  useEffect(() => {
    if (!equipmentData) {
      dispatch(getEquipment());
    } else {
      let equipments: string[] = [];
      equipmentData.forEach((equipment) => { equipments.push(equipment.name) })
      setAvailableEquipment(equipments);
    }
  }, [dispatch, equipmentData]);


  const handleSelectedCategoryItem = (filter: string) => {
    let newSelectedCategory = selectedCategory;
    if (!filter.startsWith("--")) {
      newSelectedCategory = filter;
    }
    setSelectedCategory(newSelectedCategory);
  }

  const handleSelectedEquipmentFilters = (filter: string) => {
    let newSelectedEquipment = selectedEquipment;
    if (newSelectedEquipment != filter && !filter.startsWith("--")) {
      newSelectedEquipment = filter;
    }
    setSelectedEquipment(newSelectedEquipment);
  }

  const handleSelectedOrganizations = (filter: string) => {
    let newSelectedOrganization = selectedOrganization;
    if (newSelectedOrganization != filter && !filter.startsWith("--")) {
      newSelectedOrganization = filter;
    }
    setSelectedOrganization(newSelectedOrganization);
  }

  const handleSelectIntensity = (filter: string) => {
    let newSelectedIntensityFilter = selectedIntensity;
    if (newSelectedIntensityFilter != filter && !filter.startsWith("--")) {
      newSelectedIntensityFilter = filter;
    }
    setSelectedIntensity(newSelectedIntensityFilter);
  }

  const handleSubmitKeyword = (keyword: string) => {
    let newSelectedKeyword = selectedKeyWord;
    if (newSelectedKeyword != keyword && !keyword.startsWith("--")) {
      newSelectedKeyword = keyword;
    }
    setSelectedKeyWord(newSelectedKeyword);
  }

  const handleSubmitLocation = (location: string) => {
    let newSelectedLocation = selectedIntensity;
    if (newSelectedLocation != location && !location.startsWith("--")) {
      newSelectedLocation = location;
    }
    setSelectedLocation(newSelectedLocation);
  }

  const updateSelectedFilters = () => {
    let updatedSelectedFilters: string[] = [];
    let allFilters = [selectedKeyWord, selectedLocation, selectedCategory, selectedEquipment, selectedOrganization, selectedIntensity];
    allFilters.forEach((filterCategory) => {
      if (filterCategory != "") {
        updatedSelectedFilters = updatedSelectedFilters.concat(filterCategory);
      }
    });
    setSelectedFilters(updatedSelectedFilters);
  }

  useEffect(() => {
    updateSelectedFilters();
  }, [selectedCategory, selectedEquipment, selectedIntensity, selectedKeyWord, selectedLocation, selectedOrganization])

  const handleClear = () => {
    setSelectedFilters([]);
    setSelectedCategory("");
    setSelectedEquipment("");
    setSelectedIntensity("");
    setSelectedKeyWord("");
    setSelectedOrganization("");
    setSelectedLocation("");
  }

  useEffect(() => {
    let string = "";
    // Steg1
    if (selectedKeyWord != "") {
      string += "title__icontains=" + selectedKeyWord + "&";
    }
    if (selectedOrganization != "") {
      string += "organization_owner__name__icontains=" + selectedOrganization + "&";
    }
    if (selectedLocation != "") {
      string += "location__icontains=" + selectedLocation + "&";
    }
    if (selectedCategory != "") {
      string += "categories__name__icontains=" + selectedCategory + "&";
    }
    if (selectedIntensity != "") {
      let intensity = selectedIntensity
      string += "activity_level__icontains=" + intensity.charAt(intensity.length - 1) + "&";
    }
    if (selectedEquipment != "") {
      string += "equipment_used__name__icontains=" + selectedEquipment + "&";
    }

    const url = new URL(window.location.href)
    const type = url.searchParams.get("type") //aktiviteter or arrangementer
    if (type == "aktiviteter") {
      dispatch(getEvents("user", string)); //get all activities
    } else { dispatch(getEvents("organization", string)) } // get all events
  }, [selectedFilters])

  return (
    <FilterWrapper>
      <FilterHeader> Søk etter aktiviteter: </FilterHeader>
      <SearchField tittel="søkeord" submitFunction={(keyword: string) => handleSubmitKeyword(keyword)} />
      <SearchField tittel="sted" submitFunction={(location: string) => handleSubmitLocation(location)} />
      <DropDownWrapAll>
        <DropDown
          tittel="Kategori"
          items={availableCategories}
          addFunction={(filter: string) => handleSelectedCategoryItem(filter)} />
        <DropDown
          tittel="Utstyr"
          items={availableEquipment}
          addFunction={(filter: string) => handleSelectedEquipmentFilters(filter)} />
        <DropDown
          tittel="Organisasjon"
          items={availableOrganizations}
          addFunction={(filter: string) => handleSelectedOrganizations(filter)} />
        <DropDown
          tittel="Intensistet"
          items={availableIntensity}
          addFunction={(filter: string) => handleSelectIntensity(filter)} />
      </DropDownWrapAll>
      <SelectedFilters tittel="Utvalgte filtre: " filters={selectedFilters} />
      <Button text={"Klikk for å nulstille filtre"} onClickFunc={handleClear} />
    </FilterWrapper>
  );
}

export default Filter;
