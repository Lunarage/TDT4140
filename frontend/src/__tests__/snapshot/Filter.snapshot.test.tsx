import { mount } from "enzyme";
import Enzyme from "enzyme";
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import toJSON from "enzyme-to-json";
import Filter from '../../filter/Filter';
import { store } from '../..';
import { Provider } from 'react-redux';

Enzyme.configure({ adapter: new Adapter() });

jest.mock("react-router-dom", () => ({
  ...jest.requireActual('react-router-dom') as any,
  useHistory: () => ({
    push: jest.fn(),
  }),
}));

const FilterWrapper = () => (
  <Provider store={store}>
    <Filter />
  </Provider>
);

describe("Filter", () => {
  it("renders Filter", () => {
    mount(<FilterWrapper />);
  });
  it("Should match Filter snapshot", () => {
    const comp = mount(<FilterWrapper />);
    expect(toJSON(comp)).toMatchSnapshot();
  });
});

