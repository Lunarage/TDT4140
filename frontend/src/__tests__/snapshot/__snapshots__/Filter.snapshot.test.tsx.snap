// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter Should match Filter snapshot 1`] = `
<FilterWrapper>
  <Provider
    store={
      Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      }
    }
  >
    <Filter>
      <styled.div>
        <div
          className="sc-hKgILt dVKjjV"
        >
          <styled.h2>
            <h2
              className="sc-eCssSg dLfefh"
            >
               Søk etter aktiviteter: 
            </h2>
          </styled.h2>
          <SearchField
            submitFunction={[Function]}
            tittel="søkeord"
          >
            <div>
              <form
                onSubmit={[Function]}
              >
                <label
                  className="filterSearchInputLabel"
                >
                  Skriv inn søkeord ...
                  <input
                    className="filterSearchInput"
                    onChange={[Function]}
                    type="text"
                    value=""
                  />
                </label>
              </form>
              <button
                className="inputButton"
                onClick={[Function]}
                type="submit"
              >
                Legg til søkeord
              </button>
            </div>
          </SearchField>
          <SearchField
            submitFunction={[Function]}
            tittel="sted"
          >
            <div>
              <form
                onSubmit={[Function]}
              >
                <label
                  className="filterSearchInputLabel"
                >
                  Skriv inn sted ...
                  <input
                    className="filterSearchInput"
                    onChange={[Function]}
                    type="text"
                    value=""
                  />
                </label>
              </form>
              <button
                className="inputButton"
                onClick={[Function]}
                type="submit"
              >
                Legg til sted
              </button>
            </div>
          </SearchField>
          <styled.div>
            <div
              className="sc-jSgupP ZVJFN"
            >
              <DropDown
                addFunction={[Function]}
                items={Array []}
                tittel="Kategori"
              >
                <form>
                  <label
                    className="dropDownLabel"
                  >
                    Kategori :
                  </label>
                  <select
                    className="dropDownWrapper"
                    onChange={[Function]}
                  >
                    <option
                      className="dropDownItemTitle"
                    >
                      -- Velg Kategori: --
                    </option>
                  </select>
                </form>
              </DropDown>
              <DropDown
                addFunction={[Function]}
                items={Array []}
                tittel="Utstyr"
              >
                <form>
                  <label
                    className="dropDownLabel"
                  >
                    Utstyr :
                  </label>
                  <select
                    className="dropDownWrapper"
                    onChange={[Function]}
                  >
                    <option
                      className="dropDownItemTitle"
                    >
                      -- Velg Utstyr: --
                    </option>
                  </select>
                </form>
              </DropDown>
              <DropDown
                addFunction={[Function]}
                items={Array []}
                tittel="Organisasjon"
              >
                <form>
                  <label
                    className="dropDownLabel"
                  >
                    Organisasjon :
                  </label>
                  <select
                    className="dropDownWrapper"
                    onChange={[Function]}
                  >
                    <option
                      className="dropDownItemTitle"
                    >
                      -- Velg Organisasjon: --
                    </option>
                  </select>
                </form>
              </DropDown>
              <DropDown
                addFunction={[Function]}
                items={
                  Array [
                    "Intensitet = 1",
                    "Intensitet = 2",
                    "Intensitet = 3",
                    "Intensitet = 4",
                    "Intensitet = 5",
                  ]
                }
                tittel="Intensistet"
              >
                <form>
                  <label
                    className="dropDownLabel"
                  >
                    Intensistet :
                  </label>
                  <select
                    className="dropDownWrapper"
                    onChange={[Function]}
                  >
                    <option
                      className="dropDownItemTitle"
                    >
                      -- Velg Intensistet: --
                    </option>
                    <option
                      className="dropDownItem"
                      key="Intensitet = 1"
                      value="Intensitet = 1"
                    >
                       
                      Intensitet = 1
                       
                    </option>
                    <option
                      className="dropDownItem"
                      key="Intensitet = 2"
                      value="Intensitet = 2"
                    >
                       
                      Intensitet = 2
                       
                    </option>
                    <option
                      className="dropDownItem"
                      key="Intensitet = 3"
                      value="Intensitet = 3"
                    >
                       
                      Intensitet = 3
                       
                    </option>
                    <option
                      className="dropDownItem"
                      key="Intensitet = 4"
                      value="Intensitet = 4"
                    >
                       
                      Intensitet = 4
                       
                    </option>
                    <option
                      className="dropDownItem"
                      key="Intensitet = 5"
                      value="Intensitet = 5"
                    >
                       
                      Intensitet = 5
                       
                    </option>
                  </select>
                </form>
              </DropDown>
            </div>
          </styled.div>
          <SelectedFilters
            filters={Array []}
            tittel="Utvalgte filtre: "
          >
            <styled.div>
              <div
                className="sc-gKsewC eMVLaw"
              >
                <p>
                  Utvalgte filtre: 
                </p>
              </div>
            </styled.div>
          </SelectedFilters>
          <Button
            onClickFunc={[Function]}
            text="Klikk for å nulstille filtre"
          >
            <styled.div
              onClick={[Function]}
              style={
                Object {
                  "backgroundColor": "#ff4e37",
                  "color": "white",
                  "minWidth": "8em",
                }
              }
            >
              <div
                className="sc-bdfBwQ jzdiim"
                onClick={[Function]}
                style={
                  Object {
                    "backgroundColor": "#ff4e37",
                    "color": "white",
                    "minWidth": "8em",
                  }
                }
              >
                <styled.div>
                  <div
                    className="sc-gsTCUz dfTHCv"
                  >
                    Klikk for å nulstille filtre
                  </div>
                </styled.div>
              </div>
            </styled.div>
          </Button>
        </div>
      </styled.div>
    </Filter>
  </Provider>
</FilterWrapper>
`;
